### ISSUE TITLE
Documentation on how to extract and load translations?

### ISSUE THREAD

**Issue by sontek** [i18next/i18next#995]
Hey!  We have a react application that we need to translate and are evaluating using `18next` vs `react-intl` for this.   I'm having a problem finding the documentation necessary to evaluate it and hoping you can help me!

We use the service `transifex` (https://www.transifex.com/) for all of our translations (Native apps, backend servers, etc) and want to continue using that on this react app.   What they need from us is translations extracted from our `.js` files and uploaded to them, they support many formats but we use `gettext` everywhere else ( https://docs.transifex.com/formats/gettext ).

So my question:

- How can I extract translations into a static `.json` or `.po` file that can be sent to a translation service?
- How can I use webpack to load the translated files for `i18next` to read?

I see a lot of the community is using `xhr` backend but we will not be doing this, we will be using `webpack` + `code-splitting` to make sure we only need to bundle the languages the user needs.

I did search the internet/docs before opening this issue and found this blog post which talks about exactly what I need:

https://medium.com/@jamuhl/i18n-frameworks-the-unfair-showdown-8d436cd6f470

`Round 5: How to extract translations from your code`

but the link is dead.

Ideally I could use babel/webpack to extract all the strings.

Thanks for your help!

**Comment by jamuhl** [comment]
For react and i18next have a look at react-i18next (comes with endless samples): https://github.com/i18next/react-i18next

Regarding tooling: You will find both extraction tools and bundlers/loaders for webpack here: https://www.i18next.com/plugins-and-utils.html#utils

there is also a gettext converter -> if you like to stay on gettext for transfer (on top of the utils list)

regarding transifex, not want to make advertising, but i18next has a own translation management / localization as a service tier: https://locize.com

comes with a lot of benefits and is less expensive ;) -> see the example in the react-i18next repo https://github.com/i18next/react-i18next/tree/master/example/locize-example ---> movie that says more than words https://www.youtube.com/watch?v=9NOzJhgmyQE

**Comment by sontek** [comment]
@jamuhl Thanks for the quick reply!   So I found the plugin and utils but it wasn't clear what I should be using inside there to extract translations.  Should I be using the ones marked as `extractor`?  Out of the ones on there it seems the two that might work are:

http://i18next.github.io/i18next-scanner/
https://github.com/i18next/i18next-parser

I'm not clear on when I'd use one or the other.  They both have docs for grunt and gulp but nothing about webpack or babel.

We want to avoid any package that has 1 star / 1 fork, meaning we'd be the second user of it :)

I don't specifically need to use gettext, it is just a widely supported format on translation services.  I'm willing to use any format they support.

With regards to locize, do you support uploading / downloading translation files or only fetching them via API?    We are very much against makin requests to our own backend or to others for the translations, we prefer the extract / bundle approach.   I'll definitely check it out if it supports that approach though!

**Comment by jamuhl** [comment]
Not tried them myself - but both repos are very active - so those are widely used and maintained.

In my own projects i rely on the saveMissing feature that's build in i18next backend. So everytime i18next want to translate a key and is not able to find a translation for that it submits that to the backend. This also works with locize. So best workflow would/could be using the xhr backend of locize during development and bundling for production (like eg. atroo does for its publishing tool used by major newspapers in germany and switzerland -> they use their own webpack plugin: https://github.com/atroo/i18next-resource-store-loader).

You don't have to use the translations published on locize CDN - you can download those via API or CLI (https://docs.locize.com/api.html / https://github.com/locize/locize-cli)

**Comment by jamuhl** [comment]
closing for now...feel free to reopen if still is a open question

**Comment by norayr93** [comment]
As it's more than half and year from closing date , I would like to ask whether there is any news or better approach for this case.

**Comment by jamuhl** [comment]
@norayr93 beside i18next-parser, i18next-scanner there is now also `babel-plugin-i18next-extract` https://www.i18next.com/overview/plugins-and-utils#extraction-tools

still, the most comfortable way is using the saveMissing with https://locize.com during development time.

This also gives you a proper solution at hand to involve your translators into the process 👍

**Comment by ggcaponetto** [comment]
Hello, thanks for all the valuable information regarding "key extraction". I would like to add my own open source solution to tackle this problem, to complement the other scanners and parsers mentioned above. I built this library because i had trouble with keys that use dynamic placeholders or are composed of concatenated strings. In those cases I found that LLMs help.

[https://github.com/ggcaponetto/i18next-ai-keyscan](https://github.com/ggcaponetto/i18next-ai-keyscan)
